src/uu/tac/src/tac.rs

```diff
@@ -81,6 +81,54 @@ pub fn uu_app() -> App<'static, 'static> {
         .arg(Arg::with_name(options::FILE).hidden(true).multiple(true))
 }
 
+/// Print lines of a buffer in reverse, with line separator given as a regex.
+///
+/// `data` contains the bytes of the file.
+///
+/// `pattern` is the regular expression given as a
+/// [`regex::bytes::Regex`] (not a [`regex::Regex`], since the input is
+/// given as a slice of bytes). If `before` is `true`, then each match
+/// of this pattern in `data` is interpreted as the start of a line. If
+/// `before` is `false`, then each match of this pattern is interpreted
+/// as the end of a line.
+///
+/// This function writes each line in `data` to [`std::io::Stdout`] in
+/// reverse.
+///
+/// # Errors
+///
+/// If there is a problem writing to `stdout`, then this function
+/// returns [`std::io::Error`].
+fn buffer_tac_regex(
+    data: &[u8],
+    pattern: regex::bytes::Regex,
+    before: bool,
+) -> std::io::Result<()> {
+    // Find the endpoint of each match.
+    //
+    // If the `before` flag is `true`, then a line is between the start
+    // of one match and the start of the next. Otherwise, a line is
+    // between the end of one match and the end of the next.
+    let mut endpoints: Vec<usize> = pattern
+        .find_iter(data)
```

Comment:
This is not 100% correct because the regex searches from the front while it should be searching from the back. I'm not sure how easy it is to fix, because the `regex` crate does not provide `find_reverse` or similar.
Here's an example where it matters: 
`printf "axxxb" | tac --separator="xx"` should print `baxxx`, but if you enable `--regex` it prints `xbaxx` instead.

Reply:
We basically want to search for the separator in `data[last_separator_index - 1..last_separator_index]`, then, if there's no match, search in `data[last_separator_index - 2..last_separator_index]`, and so on. This sounds like it would be very expensive, but maybe a possible solution? FWIW, this is what GNU seems to be doing.

Reply:
Good catch. I'm probably not the right person to implement a complete `find_reverse()`, but I might be able to produce a naive version that covers basic use cases like the one you just described. I will update this pull request with that change.